File: test/internal/headers_test.dart
import "dart:typed_data";
import "package:test/test.dart";
import "package:frugal/frugal.dart";

var headers = {"foo": "bar", "blah": "baz"};
var list = [0, 0, 0, 0, 29, 0, 0, 0, 3, 102, 111, 111, 0, 0, 0, 3, 98, 97,
114, 0, 0, 0, 4, 98, 108, 97, 104, 0, 0, 0, 3, 98, 97, 122];

void main() {
  test("readHeaders reads the headers out of the transport", () {
    var encodedHeaders = new Uint8List.fromList(list);
    var transport = new TUint8List(encodedHeaders);
    var decodedHeaders = readHeaders(transport);
    expect(decodedHeaders, headers);
  });

  test("readHeaders throws error for unsupported version", () {
    var encodedHeaders = new Uint8List.fromList([0x01]);
    var transport = new TUint8List(encodedHeaders);
    expect(() => readHeaders(transport), throwsUnsupportedError);
  });

  test("decodeHeadersFromFrame decodes headers from a fixed frame", () {
    var encodedHeaders = new Uint8List.fromList(list);
    var decodedHeaders = decodeHeadersFromFrame(encodedHeaders);
    expect(decodedHeaders, headers);
  });

  test("encodeHeaders encodes headers and decodeHeadersFromFrame decodes them", () {
    var encodedHeaders = encodeHeaders(headers);
    var decodedHeaders = decodeHeadersFromFrame(encodedHeaders);
    expect(decodedHeaders, headers);
  });

  test("encodeHeaders encodes null headers and decodeHeadersFromFrame decodes them", () {
    var encodedHeaders = encodeHeaders(null);
    var decodedHeaders = decodeHeadersFromFrame(encodedHeaders);
    expect(decodedHeaders, {});
  });

  test("encodeHeaders encodes empty headers and decodeHeadersFromFrame decodes them", () {
    var encodedHeaders = encodeHeaders({});
    var decodedHeaders = decodeHeadersFromFrame(encodedHeaders);
    expect(decodedHeaders, {});
  });

  test("decodeHeadersFromFrame throws error for bad frame", () {
    expect(() => decodeHeadersFromFrame(new Uint8List(3)), throwsStateError);
  });

  test("decodeHeadersFromeFrame throws error for unsupported version", () {
    var encodedHeaders = new Uint8List.fromList([0x01, 0, 0, 0, 0]);
    expect(() => decodeHeadersFromFrame(encodedHeaders), throwsUnsupportedError);
  });
}

File: lib/java/src/test/java/com/workiva/frugal/internal/HeadersTest.java
package com.workiva.frugal.internal;

import com.workiva.frugal.exception.FProtocolException;
import org.apache.thrift.TException;
import org.apache.thrift.transport.TMemoryInputTransport;
import org.apache.thrift.transport.TTransport;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;

/**
 * Tests for {@link Headers}.
 */
@RunWith(JUnit4.class)
public class HeadersTest {

    private static final Map<String, String> HEADERS;

    static {
        HEADERS = new HashMap<>();
        HEADERS.put("foo", "bar");
        HEADERS.put("blah", "baz");
    }

    private static final byte[] LIST = new byte[]{
          0, 0, 0, 0, 29, 0, 0, 0, 3, 102, 111, 111, 0, 0, 0, 3, 98, 97,
          114, 0, 0, 0, 4, 98, 108, 97, 104, 0, 0, 0, 3, 98, 97, 122
    };

    @Rule
    public final ExpectedException thrown = ExpectedException.none();

    @Test
    public void testReadOutOfTransport() throws TException {
        TTransport transport = new TMemoryInputTransport(LIST);

        Map<String, String> decodedHeaders = Headers.read(transport);
        assertEquals(HEADERS, decodedHeaders);
    }

    @Test
    public void testReadThrowsFExceptionForUnsupportedVersion() throws TException {
        TTransport transport = new TMemoryInputTransport(new byte[]{1});

        thrown.expect(TException.class);
        thrown.expectMessage("unsupported header version 1");
        Headers.read(transport);
    }

    @Test
    public void testReadThrowsTExceptionForTTransportException() throws TException {
        TTransport transport = new TMemoryInputTransport(new byte[]{0, 0, 0});

        thrown.expect(TException.class);
        // CHECKSTYLE:OFF
        thrown.expectMessage("Cannot read. Remote side has closed. Tried to read 4 bytes, but only got 2 bytes. (This is often indicative of an internal error on the server side. Please check your server logs.");
        // CHECKSTYLE:ON
        Headers.read(transport);
    }

    @Test
    public void testDecodeFromFrame() throws TException {
        Map<String, String> decodedHeaders = Headers.decodeFromFrame(LIST);
        assertEquals(HEADERS, decodedHeaders);
    }

    @Test
    public void testEncodeDecode() throws TException {
        byte[] encodedHeaders = Headers.encode(HEADERS);
        Map<String, String> decodedHeaders = Headers.decodeFromFrame(encodedHeaders);
        assertEquals(HEADERS, decodedHeaders);
    }

    @Test
    public void testEncodeDecodeNull() throws TException {
        Map<String, String> empty = new HashMap<>();
        byte[] encodedHeaders = Headers.encode(null);
        Map<String, String> decodedHeaders = Headers.decodeFromFrame(encodedHeaders);
        assertEquals(empty, decodedHeaders);
    }

    @Test
    public void testEncodeDecodeEmpty() throws TException {
        Map<String, String> empty = new HashMap<>();
        byte[] encodedHeaders = Headers.encode(empty);
        Map<String, String> decodedHeaders = Headers.decodeFromFrame(encodedHeaders);
        assertEquals(empty, decodedHeaders);
    }

    @Test
    public void testDecodeHeadersFromFrameThrowsFExceptionForBadFrame() throws TException {
        thrown.expect(FProtocolException.class);
        thrown.expectMessage("invalid frame size 3");
        Headers.decodeFromFrame(new byte[3]);
    }

    @Test
    public void testDecodeHeadersFromFrameThrowsFExceptionForUnsupportedVersion() throws TException {
        thrown.expect(FProtocolException.class);
        thrown.expectMessage("unsupported header version 1");
        Headers.decodeFromFrame(new byte[]{1, 0, 0, 0, 0});
    }

}
